pydss/flapper.py

```diff
@@ -0,0 +1,172 @@
+        self.wards[msg.sender] = 1;
+        self.vat = VatLike(vat);
+        self.gem = GemLike(gem);
+        self.live = 1;
+
+    def add(self, x, y):
+        z = x+y
+        if z > x:
+            return z
+        return -1
+
+    def mul(self, x, y):
+        z = x * y
+        if (y == 0 or (z // y == x)):
+            return z
+        return -1
+    def kick(self, lot, bid) :
+        if (self.live != 1):
+            return "Flapper/not-live"
+        if (self.kicks < -1):
+            return "Flapper/overflow";
+        sid = id(self)
+        sid += self.kicks;
+
+        self.bids[sid].bid = bid;
+        self.bids[sid].lot = lot;
+        self.bids[sid].guy = self.message.sender;
+        self.bids[sid].end = self.add(self.now, self.tau);
+
+        self.vat.move(self.msg.sender, self.address(self), lot);
+        # emit Kick(id, lot, bid);
+    """
+    // --- Admin ---
+    function file(bytes32 what, uint data) external note auth 
+        if (what == "beg") beg = data;
+        else if (what == "ttl") ttl = uint48(data);
+        else if (what == "tau") tau = uint48(data);
+        else revert("Flapper/file-unrecognized-param");
+    
+    """
+
+
+    def tick(self, sid) :
+        if self.bids[sid].end > self.now:
+            return "Flapper/not-finished"
+        if self.bids[sid].tic != 0:
+            return "Flapper/bid-already-placed"
+        self.bids[sid].end = self.add(self.now, self.tau);
+    def tend(self, sid, lot, bid):
+        if self.live != 1:
+            return "Flapper/not-live"
+        if self.bids[sid].guy == 0:
+           return "Flapper/guy-not-set";
+        if not (self.bids[sid].tic > self.now or self.bids[sid].tic == 0):
+            return "Flapper/already-finished-tic"
+        if not (self.bids[sid].end > self.now):
+            return "Flapper/already-finished-end"
+
+        if not (lot == self.bids[sid].lot):
+            return "Flapper/lot-not-matching"
+        if not (bid >  self.bids[sid].bid):
+            return "Flapper/bid-not-higher";
+        if not (self.mul(bid, self.ONE) >= self.mul(self.beg, self.bids[sid].bid)):
+            return "Flapper/insufficient-increase"
+
+        if (self.msg.sender != self.bids[sid].guy):
+            self.gem.move(self.message.sender, self.bids[sid].guy, self.bids[sid].bid);
+            self.bids[sid].guy = self.message.sender;
+
+        self.gem.move(self.msg.sender, id(self.message.sender), bid - self.bids[sid].bid);
+
+        self.bids[sid].bid = bid;
+        self.bids[sid].tic = self.add(self.now, self.ttl);
+
+    def deal( self,sid):
+        if not(self.live == 1):
+            return "Flapper/not-live"
+        if not(self.bids[sid].tic != 0 and (self.bids[sid].tic < self.now or self.bids[sid].end < self.now)):
+            return "Flapper/not-finished"
+        self.vat.move(sid, self.bids[sid].guy, self.bids[sid].lot);
+        self.gem.burn(sid, self.bids[sid].bid);
+        del self.bids[sid];
+
+
+    def cage(self, rad):
```

Comment:
we're not worrying about emergency shutdown (yet), so can remove `cage` and `yank`